Partial modification of tests

sroy4899 opened the pull request:
Modified all tests of the form `b<number>_<description` to add alltypes annotations. Decided to revert adding checked region annotations until issues surrounding it are resolved, but made modifications to scripts so that once they are, all autogenerated tests and the existing `b<number>` tests can be updated with little to no hassle.

mwhicks1 commented:
High-level question: Did you trust that the tool "knew what it was doing" when updating these files to use `-alltypes` or did you scrutinize the output to see whether it made sense?

mwhicks1 commented on clang/test/CheckedCRewriter/processor.py:
```diff
@@ -0,0 +1,187 @@
+import fileinput 
```
This is the script that automatically converts the tests? If you used it, did you check that the output actually makes sense? I.e., not just trust the output of the tool to be right, but confirm that it makes sense, by hand?

mwhicks1 commented on clang/test/CheckedCRewriter/testgenerator.py:
```diff
@@ -484,7 +484,7 @@ def process_file(file, cname, cname2, proto, alltypes, new_defs, new_defs2, susp
 
     keywords = "int char struct double float".split(" ") 
-    ckeywords = "_Ptr _Array_ptr _Nt_array_ptr".split(" ")
+    ckeywords = "_Ptr _Array_ptr _Nt_array_ptr _Checked _Unchecked".split(" ")
```
In your comment at the start of this PR you said you had backed out the changes that added checked regions, but it seems that the script still adds them?

sroy4899 replied:
I have these added so when the checked regions issues get resolved, I can just run the script without changes. None of the tests have been updated with these keywords in them yet :)

sroy4899 commented:
I auto-gen'd these annotations so I blindly trusted the tool on this one. I scrutinized most of the results for the manual test update PR.

mwhicks1 commented on clang/test/CheckedCRewriter/b9_allsafestructp.c:
```diff
@@ -17,29 +22,35 @@ struct p {
   int *x;
   char *y;
 };
-//CHECK: _Ptr<int> x;
```
Why have these gone away? Seems like all of the `CHECK` annotations you had before should be `CHECK_NOALL` now, right?

sroy4899 replied:
Ah, I added annotations almost purely to the functions as opposed to fields of structs I used. Thanks for pointing this one out. I can change the tests to reflect this!

mwhicks1 commented on clang/test/CheckedCRewriter/b10_allsafepointerstruct.c:
```diff
@@ -28,7 +33,11 @@ struct p *sus(struct p *x, struct p *y) {
 }
-//CHECK: _Ptr<struct p> sus(_Ptr<struct p> x, _Ptr<struct p> y) {
+//CHECK_NOALL: _Ptr<struct p> sus(_Ptr<struct p> x, _Ptr<struct p> y) {
+//CHECK_NOALL:   _Ptr<struct p> z =  malloc<struct p>(sizeof(struct p));
```
Glad that the `malloc` instantiations got added.

mwhicks1 commented on clang/test/CheckedCRewriter/b10_allsafepointerstruct.c:
```diff
@@ -40,7 +49,15 @@ struct p *foo() {
+//CHECK_NOALL:  _Ptr<struct p> x = ((void *)0);
+//CHECK_NOALL: _Ptr<struct p> y = ((void *)0);
+//CHECK_NOALL:   _Ptr<struct p> z =  (struct p *) sus(x, y);
+//CHECK_ALL: _Ptr<struct p> foo(void) {
```
Have you confirmed that all the extra `CHECK_NOALL`s are reasonable, i.e., that the tool is doing the right thing?

sroy4899 replied:
Nope, but I'm going to be revisiting this batch of tests in order to add checked annotations to the structs, so I can take a look to make sure the output makes sense.

mwhicks1 replied:
Same comment as above about the annotation location.

If this is going to be a big pain to do, then we could put it off as an enhancement and just take things as they are.

sroy4899 commented:
Ok, added struct annotations, cleaned up the compilation phase of the tests to be one line instead of three lines (as discussed in #186), and combed through about ~1/4 of them to make sure they're right. The extra annotations looked concerning at first glance because I added annotations very sporadically before whereas now, every pointer (converted or not), gets annotated. In any case, the annotations make sense and seem to be doing the right thing.

mwhicks1 commented on clang/test/CheckedCRewriter/b10_allsafepointerstruct.c:
```diff
@@ -23,12 +26,24 @@ struct r {
+//CHECK_ALL:   _Ptr<int> x;
+//CHECK_ALL:   char *y;
+//CHECK_ALL:   _Ptr<struct r> next;
+
```
I'm sorry to ask, but is it possible for these annotations to be added closer to the variables in question? I'd prefer to see
```
struct p {
  int *x;
//CHECK_NOALL:   _Ptr<int> x;
//CHECK_ALL:   _Ptr<int> x;
  char *y;
//CHECK_NOALL:   char *y;
//CHECK_ALL:   char *y;
};
struct r {
  int data;
  struct r *next;
//CHECK_NOALL:   _Ptr<struct r> next;
//CHECK_ALL:   _Ptr<struct r> next;
};

mwhicks1 replied:
For that matter: If you can make a single annotation `CHECK: ...` when the two are the same, so much the better. You had this with other tests previously, right?

Doing these things will make the tests easier to read and understand.

sroy4899 replied:
This makes sense to me. It'll be a bit more complex but I think I can pull it off. 👍

mwhicks1 replied:
Looking over all of the output, it seems that the differences between using `-alltypes` and not using it are very slight. This makes me more inclined to say that it would be great to merge the annotations. I.e., for the above:

```
// RUN: cconv-standalone -alltypes %s -- | FileCheck -match-full-lines -check-prefixes="CHECK" %s
//RUN: cconv-standalone %s -- | FileCheck -match-full-lines -check-prefixes="CHECK" %s
// RUN: cconv-standalone %s -- | %clang -c -fcheckedc-extension -x c -o /dev/null -
...
struct p {
  int *x;
//CHECK:   _Ptr<int> x;
  char *y;
//CHECK:   char *y;
};
struct r {
  int data;
  struct r *next;
//CHECK:   _Ptr<struct r> next;
};
```
I think this is a lot easier to read.

mwhicks1 commented on clang/test/CheckedCRewriter/b11_calleestructnp.c:
```diff
@@ -53,3 +72,9 @@ struct np *bar() {
 //struct np *bar() {
+//CHECK_NOALL: struct np *bar() {
+//CHECK_NOALL:   struct np *z = sus(x, y);
+//CHECK_NOALL: //struct np *bar() {
```
Why is this `//struct np *bar()` ? Seems odd. Maybe we should just delete the commented-out code that's there now?

mwhicks1 commented on clang/test/CheckedCRewriter/b24_implicitunsafecast.c:
```diff
@@ -32,4 +49,13 @@ char* bar() {
+//CHECK_NOALL: char* bar() {
+//CHECK_NOALL: int *x = &sx;
+//CHECK_NOALL: _Ptr<int> y = &sy;
+//CHECK_NOALL:   // Because, sus returns int* but assigning to char *
```
Not necessary?

sroy4899 commented:
Ok, did a lot of refactoring and cleaning, and I think I've got it to a point where I'm happy when I look at the new tests.

sroy4899 commented:
Actually, please hold off on merging (if everything's fine). I need to check that all the tests pass first!

sroy4899 commented:
TL;DR: Discovered a frustrating bug, thought I'd post a quick status to explain the holdup on what should've been a pretty cut-and-dry modification.

Currently, my script generates annotations like this: 
``` 
1      struct p *foo() {
2           int ex1 = 2, ex2 = 3;
3           struct p *x, *y;
4           x->x = &ex1;
5           y->x = &ex2;
6           x->y = &ex2;
7           y->y = &ex1;
8          struct p *z = (struct p *) sus(x, y);
9          return z;
10    } 

11  //CHECK_NOALL: _Ptr<struct p> foo(void) {
12  //CHECK_NOALL: _Ptr<struct p> z =  (struct p *) sus(x, y);
13  //CHECK_ALL: struct p * foo(void) {
14  //CHECK: _Ptr<struct p> x = ((void *)0);
15  //CHECK: _Ptr<struct p> y = ((void *)0);
16  //CHECK_ALL: struct p *z = (struct p *) sus(x, y);
``` 
Like we discussed, CHECK represents things that are common between alltypes and non-alltypes. However, FileCheck rejects this file because on line 12, CHECK_NOALL annotates the pointer `z`, which occurs in the source code on line 8. However, when it reaches line 14, it is trying to check the pointer `x`, which occurs in the source code on line 3! It can't jump backwards like this. If I were to put the CHECK before (by swapping , an equivalent problem would be presented but in this case for CHECK_ALL (because CHECK_ALL jumps down to line 3 in the source code to annotate `x` and then tries to jump up to annotate the function declaration on line 1). 
Several ideas on how to fix it fell through, but I've got a couple more I'm trying now.

mwhicks1 commented:
> Currently, my script generates annotations like this:

What happens if you try to put your annotations closer to the lines that are changed, rather than grouped at the end? That was a "feature request" from me. Might it help you solve this problem?

sroy4899 commented:
I tried a version of that, where I generate the alltypes and non-alltypes version and traverse them line by line, but the line lengths differ (sometimes) based on whether we enable alltypes or not (see example below) so it's not as robust a solution as I would like (though it might be my fallback in case my next idea doesn't pan out). 
``` 
//SOURCE
int *x, *y;      // pretend x and y are both array ptrs 

//alltypes 
Array_ptr<int> x = ... 
Array_ptr<int> y = ... 

//non-alltypes 
int *x, *y;
```

mwhicks1 commented:
Maybe change the source file to put the definitions like these on different lines?

You can make separate tests that ensure that multiple definitions are split across lines.

sroy4899 commented:
The annotations have been unified for all and noall when they're the same, annotations occur right after the definition, and all the tests pass. The script is also much cleaner. 😄

mwhicks1 commented on clang/test/CheckedCRewriter/b12_callerstructnp.c:
```diff
@@ -8,48 +10,61 @@ extern _Itype_for_any(T) void *realloc(void *pointer : itype(_Array_ptr<T>) byte
+
 struct np *sus(struct p x, struct p y) {
+	//CHECK_NOALL: struct np *sus(struct p x, struct p y) : itype(_Ptr<struct np>) {
+	//CHECK_ALL: struct np * sus(struct p x, struct p y) {
```
This is very strange: Enabling `-alltypes` results in a WILD pointer whereas not enabling it results in a checked pointer. I don't know why that would happen. We can accept this PR, but perhaps put this on your list to investigate?

mwhicks1 replied:
Might have something to do with how `malloc` is handled in the two cases.

sroy4899 replied:
This is weird. Looking into it now

mwhicks1 commented on clang/test/CheckedCRewriter/b29_structprotocastsafeuseunsafe.c:
```diff
@@ -8,53 +10,72 @@ extern _Itype_for_any(T) void *realloc(void *pointer : itype(_Array_ptr<T>) byte
 struct r *foo() {
-  struct r *x, *y;
+	//CHECK_NOALL: _Ptr<struct r> foo(void) {
+	//CHECK_ALL: struct r *foo() {
```
Same surprising behavior as the other test I commented on.

sroy4899 commented:
Commenting here as opposed to separately on the two tests. I think the fault here isn't with how the tool is converting, but the test itself. In both (or perhaps more) of the tests with this weird behavior, `foo()` uses `sus()` exactly how we expect: call the function, return it, done. `bar()` calls the function and then adds 2 to it. This is legacy code where adding 2 was supposed to be an "unsafe" use (back when these tests weren't run with alltypes). As you can imagine, with alltypes, adding 2 is an indication to the tool that `z` is an _Array_ptr. With alltypes on, what I think is happening is that the tool smartly sees the discrepancy of returning a normal ptr vs using it as an array in `bar()` and makes it WILD. Without alltypes, the `z+=2` is marked WILD and the tool concludes that the function `sus` returns a _Ptr for all intents and purposes. 

The solution here is to take away the ungainly `z+=2` annotations and instead do something else unsafe like `z = (type *) 5`
